src/libraries/System.Private.CoreLib/src/System/SearchValues/Strings/Helpers/StringSearchValuesHelper.cs

```diff
@@ -0,0 +1,294 @@
+            public static char TransformInput(char input) => input;
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public static Vector128<byte> TransformInput(Vector128<byte> input) => input;
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public static Vector256<byte> TransformInput(Vector256<byte> input) => input;
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public static Vector512<byte> TransformInput(Vector512<byte> input) => input;
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public static bool Equals(ref char matchStart, string candidate)
+            {
+                ref char end = ref Unsafe.Add(ref matchStart, candidate.Length);
+                ref char candidateRef = ref Unsafe.AsRef(candidate.GetPinnableReference());
+
+                do
+                {
+                    if (candidateRef != matchStart)
+                    {
+                        return false;
+                    }
+
+                    matchStart = ref Unsafe.Add(ref matchStart, 1);
+                    candidateRef = ref Unsafe.Add(ref candidateRef, 1);
+                }
+                while (Unsafe.IsAddressLessThan(ref matchStart, ref end));
+
+                return true;
+            }
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public static bool Equals<TValueLength>(ref char matchStart, string candidate)
+                where TValueLength : struct, IValueLength
+            {
+                Debug.Assert(candidate.Length > 1);
+                Debug.Assert(matchStart == candidate[0], "This should only be called after the first character has been checked");
+
+                ref byte first = ref Unsafe.As<char, byte>(ref matchStart);
+                ref byte second = ref Unsafe.As<char, byte>(ref candidate.GetRawStringData());
+                nuint byteLength = (nuint)(uint)candidate.Length * 2;
+
+                if (TValueLength.AtLeast8Chars)
+                {
+                    return SpanHelpers.SequenceEqual(ref first, ref second, byteLength);
+                }
+                else if (TValueLength.AtLeast4Chars)
+                {
+                    nuint offset = byteLength - sizeof(ulong);
+                    ulong differentBits = Unsafe.ReadUnaligned<ulong>(ref first) - Unsafe.ReadUnaligned<ulong>(ref second);
+                    differentBits |= Unsafe.ReadUnaligned<ulong>(ref Unsafe.Add(ref first, offset)) - Unsafe.ReadUnaligned<ulong>(ref Unsafe.Add(ref second, offset));
+                    return differentBits == 0;
+                }
+                else
+                {
+                    nuint offset = byteLength - sizeof(uint);
+
+                    return Unsafe.ReadUnaligned<uint>(ref Unsafe.Add(ref first, offset))
+                        == Unsafe.ReadUnaligned<uint>(ref Unsafe.Add(ref second, offset));
```

Comment:
Couldn't we be here for candidate.Length == 3? In which case, aren't we possibly missing comparing one character?

Reply:
Oh, nevermind, I see the assert at the beginning of the method saying we've already checked the first character.